gpu/impl/SimpleIMICut.cu

```diff
@@ -0,0 +1,268 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD+Patents license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+
+#include "SimpleIMICut.cuh"
+#include "../../FaissAssert.h"
+#include "../utils/ConversionOperators.cuh"
+#include "../utils/DeviceDefs.cuh"
+#include "../utils/DeviceUtils.h"
+#include "../utils/Float16.cuh"
+#include "../utils/MathOperators.cuh"
+#include "../utils/PtxUtils.cuh"
+#include "../utils/StaticUtils.h"
+#include "../utils/Reductions.cuh"
+#include "math.h"
+
+namespace faiss { namespace gpu {
+
+// Input: (batch x dim), # repeats
+// Output: (# repeats, norm of batch vector)
+// Done under the presumption that the dimension size is not too large
+// (<10k or so), since there wouldn't be enough parallelism applying a
+// single block to the problem. Also that each vector is large enough
+// (>64), since a single block works on multiple rows' norms at the
+// same time.
+// T: the type we are doing the math in (e.g., float, half)
+// TVec: the potentially vectorized type we are loading in (e.g.,
+// float4, half2)
+template <typename T, typename TVec, typename int64_t,
+          int RowTileSize, bool NormLoop>
+__global__ void SimpleIMICut(Tensor<TVec, 3, true, int64_t> input,
+                                Tensor<int, 2, true, int64_t> output
+                                int squareLen,
+                                int totalLen) {
+  extern __shared__ char smemByte1[]; // #warps * RowTileSize elements
+  extern __shared__ char smemByte2[];
+  T* smemMin = (T*) smemByte1;
+  T* smemId  = (T*) smemByte2;
+  int64_t numWarps = utils::divUp(blockDim.x, kWarpSize);
+  int64_t laneId = getLaneId();
+  int64_t warpId = threadIdx.x / kWarpSize;
+
+  bool lastRowTile = (blockIdx.x == (gridDim.x - 1));
+  int64_t rowStart = RowTileSize * blockIdx.x;
+  T rowMin[RowTileSize];
+  int rowId[RowTileSize];
+
+  if (lastRowTile) {
+    // We are handling the very end of the input matrix rows
+    for (int64_t row = 0; row < input.getSize(1) - rowStart; ++row) {
+      if (NormLoop) {
+        rowMin[0] = 1e10;
+
+        for (int64_t col = threadIdx.x;
+             col < totalLen - 2 * squareLen ; col += blockDim.x) {
+          TVec val = abs(input[0][rowStart + row][col + squareLen] - input[1][rowStart + row][col + squareLen]);
+          //rowNorm[0] = Math<T>::add(rowNorm[0], Math<TVec>::reduceAdd(val));
+          rowId[0] = ( val < rowMin[0] ? col + squareLen : rowId[0]);
```

Comment:
This is wrong. Note that `val` is a vector of four numbers. Similar lines below are also wrong.